### Business Analytics - Assignment 1 {-}


---

**Assignment Points**: 100  
**Due Date**: Friday Week 4 (17 March) @ 11.59pm  
**Submission**: Submit your file using the submission link on iLearn


- Put **all your work** into this file;
- Failure to supply solutions in the cells provided below each question will result in a loss of 30 points;
- Follow all instructions closely and **do not** print your variables to the screen unless explicitly asked to do so;
    - Comment out print statements where necessary and make sure that your submitted notebook does not have the output of previously executed print statements;
    - 10 marks will be deducted for every redundant print statement not explicitly asked for.



### About the Assignment

Credit score cards are used as a risk control method in the financial industry. Personal information submitted by credit card applicants are used to predict the probability of future defaults. The bank employs such data to decide whether to issue a credit card to the applicant or not.




| Feature Name         | Explanation     | Additional Remarks |
|--------------|-----------|-----------|
| ID | Randomly allocated client number      |         |
| Income   | Annual income  |  |
| Gender   | Applicant's Gender   | Male = 0, Female = 1  |
| Car | Car Ownership | Yes = 1, No = 0 | 
| Children | Number of Children | |
| Real Estate | Real Estate Ownership | Yes = 1, No = 0 
| Days Since Birth | No. of Days | Count backwards from current day (0), -1 means yesterday
| Days Employed | No. of Days | Count backwards from current day(0). If positive, it means the person is currently unemployed.
| Payment Default | Whether a client has overdue credit card payments | Yes = 1, No = 0



---


### Problem 1 - (50 points) {-}


**Question 1** 

- Import the `assignment_data.xlsx` file from `data` folder into a pandas DataFrame named `df`; 
- Delete duplicate rows from `df` according to `ID`;
- Delete the `ID` column.
- How many rows are left in `df`?

(10 points)

In [54]:
import pandas as pd
df = pd.read_excel('assignment_data.xlsx')
df = df.drop_duplicates(subset = 'ID', keep="first") #DF2 is dataframe without duplicates
df = df.drop(['ID'], axis = 1)
df.shape
df

Gender  Car  Real Estate  Children    Income  Days Since Birth  \
0          0    1            0         0  135000.0            -10239   
1          0    0            1         0  202500.0            -13645   
2          1    0            1         0  157500.0            -16628   
3          0    0            0         1  270000.0            -10108   
4          1    1            1         1  135000.0            -15481   
...      ...  ...          ...       ...       ...               ...   
9992       0    1            0         0  225000.0             -9479   
9993       0    1            1         2  211500.0            -10435   
9994       1    0            0         0   76500.0            -16876   
9995       0    1            1         0  180000.0            -17115   
9999       1    0            1         0  202500.0            -15047   

      Days Employed  Payment Default  
0            365243                1  
1             -5374                1  
2             -3654                1  
3             -1473                1  
4             -2874                0  
...             ...              ...  
9992          -1143                1  
9993          -2987                1  
9994            -97                0  
9995          -4599                0  
9999          -1720                1  

[5976 rows x 8 columns]

There are 5976 rows left in the Dataframe after dropping duplicate ID's.


---

**Question 2**

- Reset the index in `df` using an appropriate function from `pandas` so that the new index corresponds to the number of rows (make sure to delete the old index). 
- How many positive values of `Days Employed` are there?
- Replace the positive values of `Days Employed` with 0 (zero) in `df`

(10 points)

In [55]:
ResetDF = df.reset_index()
ResetDF = ResetDF.drop(['index'], axis = 1)
sum(ResetDF['Days Employed']>0)
ResetDF['Days Employed'] = ResetDF['Days Employed'].apply(lambda x: 0 if x>0 else x)
ResetDF

Gender  Car  Real Estate  Children    Income  Days Since Birth  \
0          0    1            0         0  135000.0            -10239   
1          0    0            1         0  202500.0            -13645   
2          1    0            1         0  157500.0            -16628   
3          0    0            0         1  270000.0            -10108   
4          1    1            1         1  135000.0            -15481   
...      ...  ...          ...       ...       ...               ...   
5971       0    1            0         0  225000.0             -9479   
5972       0    1            1         2  211500.0            -10435   
5973       1    0            0         0   76500.0            -16876   
5974       0    1            1         0  180000.0            -17115   
5975       1    0            1         0  202500.0            -15047   

      Days Employed  Payment Default  
0                 0                1  
1             -5374                1  
2             -3654                1  
3             -1473                1  
4             -2874                0  
...             ...              ...  
5971          -1143                1  
5972          -2987                1  
5973            -97                0  
5974          -4599                0  
5975          -1720                1  

[5976 rows x 8 columns]

There were 967 positive values for Days Employed, meaning there were 967 instances of unemployment.

---
**Question 3**

Create two new variables in `df` named 

1. `Age`;
2. `Years in Employment`,

which measure age and employment length in **years** (decimal numbers) from `Days Since Birth` and `Days Employed` by applying approapriate transformations on these variables. 

Delete the original variables `Days Since Birth` and `Days Employed`.

(5 points)


In [56]:
ResetDF['Age'] = ResetDF.apply(lambda x: round(abs(x['Days Since Birth'])/365, 3), axis = 1)
ResetDF['Years in Employment'] = ResetDF.apply(lambda x: round(abs(x['Days Employed'])/365,3), axis = 1)
ResetDF=ResetDF.drop(['Days Since Birth'], axis = 1)
ResetDF=ResetDF.drop(['Days Employed'], axis = 1)

In [58]:
ResetDF

Gender  Car  Real Estate  Children    Income  Payment Default     Age  \
0          0    1            0         0  135000.0                1  28.052   
1          0    0            1         0  202500.0                1  37.384   
2          1    0            1         0  157500.0                1  45.556   
3          0    0            0         1  270000.0                1  27.693   
4          1    1            1         1  135000.0                0  42.414   
...      ...  ...          ...       ...       ...              ...     ...   
5971       0    1            0         0  225000.0                1  25.970   
5972       0    1            1         2  211500.0                1  28.589   
5973       1    0            0         0   76500.0                0  46.236   
5974       0    1            1         0  180000.0                0  46.890   
5975       1    0            1         0  202500.0                1  41.225   

      Years in Employment  
0                   0.000  
1                  14.723  
2                  10.011  
3                   4.036  
4                   7.874  
...                   ...  
5971                3.132  
5972                8.184  
5973                0.266  
5974               12.600  
5975                4.712  

[5976 rows x 8 columns]

---
**Question 4**

- Create a **one**-dimensional NumPy array named `y` by exporting the first 5,000 observations of `Payment_Default`. (Hint: see `ravel()` function)
- Create a NumPy array named `X` by exporting the first 5,000 observations of the following columns `Gender`, `Car`, `Real Estate`, `Children`, `Income`, `Age`, `Years in Employment`.
 
(10 points)


In [91]:
import numpy as np
y = np.ravel(ResetDF.loc[:4999]['Payment Default'], order = 'C')
X = ResetDF.loc[:4999][['Gender','Car','Real Estate','Children','Income','Age','Years in Employment']].to_numpy()
X.shape
X

array([[0.0000e+00, 1.0000e+00, 0.0000e+00, ..., 1.3500e+05, 2.8052e+01,
        0.0000e+00],
       [0.0000e+00, 0.0000e+00, 1.0000e+00, ..., 2.0250e+05, 3.7384e+01,
        1.4723e+01],
       [1.0000e+00, 0.0000e+00, 1.0000e+00, ..., 1.5750e+05, 4.5556e+01,
        1.0011e+01],
       ...,
       [0.0000e+00, 1.0000e+00, 1.0000e+00, ..., 2.2500e+05, 5.6682e+01,
        0.0000e+00],
       [1.0000e+00, 1.0000e+00, 1.0000e+00, ..., 9.0000e+04, 3.0575e+01,
        5.2160e+00],
       [1.0000e+00, 0.0000e+00, 1.0000e+00, ..., 1.3500e+05, 5.6855e+01,
        0.0000e+00]])

---

**Question 5** 

- Use an appropriate `scikit-learn` library we learned in class to create the following NumPy arrays: `y_train`, `y_test`, `X_train` and `X_test` by splitting the data into 70% training and 30% test datasets. 
- Set `random_state` to 0 and stratify subsamples so that train and test datasets have roughly equal proportions of the target's class labels. 

(5 points) 

In [90]:
# add your code here
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 0, stratify = y)
# print(X_train.shape)
# print(y_train.shape)
# print(X_test.shape)
# print(y_test.shape)
print(y_train)
print(y_test)

[1 0 0 ... 1 1 0]
[0 0 0 ... 0 0 0]


---

**Question 6**

- Create new variables by using an appropriate `scikit-learn` library we learned in class to standardize the features from the training and test datasets to mean zero and variance one. Name the new variables by appending '_scaled' to the original variable names.


(10 points)   

In [131]:
# add your code here

---

## Problem 2 - (20 Points) {-}

**Question 7**

Fit the following two classifiers to the transformed training dataset using `scikit-learn` libraries.

- Perceptron - name your instance `pc` set `random_state=1`
- Logistic Regression - name your instance `lr` set `random_state=1`

When initializing instances of the above classifiers only set the parameters referenced above and nothing else.

(20 points)

In [132]:
# add your code here

---
## Problem 3 - (30 points) {-}


**Question 8**

- Using a method built into each of the two classifiers compute their prediction accuracies on the training data;
- Store the accuracy values into variables named according to the following pattern: `classifier_name_accuracy_train`, e.g. you should have `lr_accuracy_train`; 
- Print the two accuracy **variables** along with their brief descriptions.

(10 points)

In [133]:
# add your code here

---

**Question 9** 

- Using a method built into each of the above classifiers compute their prediction accuracy for the test dataset
- Store the accuracy values into variables named according to the following pattern: `classifier_name_accuracy_test`, e.g. you should have `lr_accuracy_test`. 
- Print the two accuracy **variables** along with brief descriptions.

(10 points)

In [134]:
# add your code here

---

**Question 10** 

Using nicely formated text in Markdown comment on the accuracies computed in Questions 8 & 9 making sure you address:
- training and test set datasets; 
- Perceptrion and Logistic Regression models. 

Are the results as expected, and why or why not? (Hint: You are not expected to comment on why a particular model is better.) 

(10 marks)


--- add your text answer here